### PR TITLE
1614 Soft deletes for master list

### LINKED ISSUE #1614: Soft deletes for master list
## Is your feature request related to a problem? Please describe 👀
Transferred from #1584 
Currently, central server allows master lists to be deleted if the master list is not linked to any stores. i.e., If a user removes the master list from all stores, it could result in an error.

### Central Changes
- Master lists shouldn't be deleted from central as program uses master list as a FK, and it will cause errors if a master list with a program is deleted
- Have a `is_active` boolean to toggle if the master list is active or not instead of deleting (central can use this to hide master lists for places that require master list deletions)
### Remote Changes
- Add is_active field to master list and filter by this field wherever applicable
- Add is_active field to program table

### Additional context 💌
- Needs issue in mSupply as well

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #1614

# 👩🏻‍💻 What does this PR do? 
Removes hard-coded missing master list and adds `is_active` field to master_list. 

# 🧪 How has/should this change been tested? 
- [ ] Check tests are still running
- [ ] Delete a master list in central
- [ ] Sync
- [ ] See that master list will now have is_active = false

### CODE REVIEW

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
Some questions:
- In the repositories, should we add a new default filter that excludes master_lists and programs that are not active?
- For existing DB we don't fix the masterlists with MISSING_PROGRAM id, do we? probably not an issue to leave them, right?

One change request: Please remove the now unused MISSING_PROGRAM constant and fix the resulting compile error in the translation :)

**Comment by roxy-dao** [comment]
> Some questions:
> 
> * In the repositories, should we add a new default filter that excludes master_lists and programs that are not active?
> * For existing DB we don't fix the masterlists with MISSING_PROGRAM id, do we? probably not an issue to leave them, right?
> 
> One change request: Please remove the now unused MISSING_PROGRAM constant and fix the resulting compile error in the translation :)

1. Don't need a filter for it since you can only delete master lists on central if they aren't linked to a store... so it shouldn't show up on any master list list anyway.
2. Ah, you're right. Will add the missing_program back in for existing dbs.

**Comment by clemens-msupply** [comment]
>     1. Don't need a filter for it since you can only delete master lists on central if they aren't linked to a store... so it shouldn't show up on any master list list anyway.

I meant isn't it possible that you have a master_list + master_list_name_join synced, then the master_list_name_join gets deleted, and then the master list is deleted as well. This means there is an inactive master list on the remote server. This master list could be queried, can't it? Or is this not possible on to do?

If not possible could you please add a comment in MasterListRepository on why we don't have to filter out in-active master_lists?

**Comment by clemens-msupply** [comment]
>     2. Ah, you're right. Will add the missing_program back in for existing dbs. Though, only DJ is using programs, so might not be an issue? Idk

You mean replacing `missing_program` references in existing rows? Yes I don't think its a problem to leave these in (?) Or is it possible to undelete a master list? If it is a problem it would be easier and safer to trigger a re-integration of the affected tables, but we haven't implemented this yet, right?

**Comment by roxy-dao** [comment]
> > ```
> > 1. Don't need a filter for it since you can only delete master lists on central if they aren't linked to a store... so it shouldn't show up on any master list list anyway.
> > ```
> 
> I meant isn't it possible that you have a master_list + master_list_name_join synced, then the master_list_name_join gets deleted, and then the master list is deleted as well. This means there is an inactive master list on the remote server. This master list could be queried, can't it? Or is this not possible on to do?
> 
> If not possible could you please add a comment in MasterListRepository on why we don't have to filter out in-active master_lists?

Why would we want to query a master list that has been deleted? Are we thinking of reinstating deleted master lists in the future? (don't think mSupply will be allowing this, so don't know if we want to deviate?)

Only having it soft delete because it causes issues for program-related master lists... (i.e. foreign key constraint...)

File: server/repository/src/db_diesel/master_list.rs [hunk 1]
```diff
@@ -65,6 +65,9 @@ impl<'a> MasterListRepository<'a> {
         let mut query = master_list_dsl::master_list.into_boxed();
 
         if let Some(f) = filter {
+            // Filter out inactive master lists by default
+            query = query.filter(master_list_dsl::is_active.eq(true));
```
Comment (clemens-msupply) [inline server/repository/src/db_diesel/master_list.rs:69]: We have to always do it not only if the user specified a filter...

File: server/repository/src/migrations/v1_06_00/mod.rs [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+
 mod contact_trace;
+mod master_list;
 
 pub(crate) struct V1_06_00;
 
 impl Migration for V1_06_00 {
     fn version(&self) -> Version {
-        Version::from_str("1.6.0")
+        Version::from_str("1.6.00")
```
Comment (clemens-msupply) [inline server/repository/src/migrations/v1_06_00/mod.rs:12]: Is there a difference between "1.6.0" and "1.6.00" or does this results in the same `Version`? so this is just a style fix?

Comment (roxy-dao) [reply server/repository/src/migrations/v1_06_00/mod.rs]: Shouldn't be any difference... have changed back to .0 to match rest of the migrations

File: server/repository/src/db_diesel/master_list.rs (hunk 1 above)
Comment (roxy-dao) [reply server/repository/src/db_diesel/master_list.rs]: [c3d8473](https://github.com/msupply-foundation/open-msupply/pull/2425/commits/c3d8473b7168b208651d37d22d5de5e273cb072f)

**Review by clemens-msupply — APPROVED** [review]
thanks!

**Pull request merged**